### PR TITLE
added crud for questions

### PR DESCRIPTION
**Pull request by MDGcode**

### CODE REVIEW

File: server/app/index.ts [hunk 1]
```diff
@@ -1,12 +1,17 @@
 import express, { Express, Request, Response } from "express";
 import serverless from "serverless-http";
 import dotenv from "dotenv";
+import cors from "cors";
+import { questionRouter } from "./question/question.router";
```
Comment (cristim67) [inline server/app/index.ts:5]: move `question` folder to `../services/question` new folder.

File: server/app/question/question.router.ts [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+import type { Request, Response } from "express";
+import { body, validationResult } from "express-validator";
+
+import * as QuestionService from "./question.service";
+import { request } from "http";
+
+export const questionRouter = express.Router();
+
+//GET all questions
+questionRouter.get("/", async (req: Request, res: Response) => {
```
Comment (cristim67) [inline server/app/question/question.router.ts:11]: declare new function `getQuestion` that have this code and import from `services`.

the code should be like that:

`questionRouter.get("/", getQuestion)`

File: server/app/question/question.router.ts [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+  try {
+    const questions = await QuestionService.listQuestions();
+    return res.status(200).json(questions);
+  } catch (error: any) {
+    return res.status(500).json(error.message);
+  }
```
Comment (cristim67) [inline server/app/question/question.router.ts:17]: console.log the error message.

File: server/app/question/question.router.ts [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+
+import * as QuestionService from "./question.service";
+import { request } from "http";
+
+export const questionRouter = express.Router();
+
+//GET all questions
+questionRouter.get("/", async (req: Request, res: Response) => {
+  try {
+    const questions = await QuestionService.listQuestions();
```
Comment (cristim67) [inline server/app/question/question.router.ts:13]: check is null.

File: server/app/question/question.router.ts [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+questionRouter.get("/:id", async (req: Request, res: Response) => {
+  const id: number = parseInt(req.params.id, 10);
+  try {
+    const question = await QuestionService.getQuestion(id);
+    if (question) return res.status(200).json(question);
+    return res.status(404).json({
+      errors: "Author could not be found",
+    });
+  } catch (error: any) {
+    return res.status(500).json(error.message);
```
Comment (cristim67) [inline server/app/question/question.router.ts:30]: same here.

File: server/app/question/question.router.ts [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+  try {
+    const questions = await QuestionService.listQuestions();
+    return res.status(200).json(questions);
+  } catch (error: any) {
+    return res.status(500).json(error.message);
+  }
+});
+
+//GET by id
+questionRouter.get("/:id", async (req: Request, res: Response) => {
```
Comment (cristim67) [inline server/app/question/question.router.ts:21]: same here.

File: server/app/question/question.router.ts [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+  const id: number = parseInt(req.params.id, 10);
+  try {
+    const question = await QuestionService.getQuestion(id);
```
Comment (cristim67) [inline server/app/question/question.router.ts:24]: here you need the type for `question`. create an interface for that.

File: server/app/question/question.router.ts [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+    if (question) return res.status(200).json(question);
+    return res.status(404).json({
+      errors: "Author could not be found",
+    });
```
Comment (cristim67) [inline server/app/question/question.router.ts:28]: reverse this, always check is wrong first. ` if (!question)`

File: server/app/question/question.router.ts [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+    if (!errors.isEmpty()) {
+      return response.status(400).json({ errors: errors.array });
+    }
+    try {
+      const question = req.body;
+      const newQuestion = await QuestionService.createQuestion(question);
+      return res.status(201).json(newQuestion);
+    } catch (error: any) {
+      return res.status(500).json({ errors: error.message });
+    }
```
Comment (cristim67) [inline server/app/question/question.router.ts:49]: move the try catch block to group all your code in this function.

File: server/app/question/question.router.ts [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+questionRouter.post(
+  "/",
+  body("question").isString(),
+  async (req: Request, res: Response) => {
+    const errors = validationResult(request);
+    if (!errors.isEmpty()) {
+      return response.status(400).json({ errors: errors.array });
+    }
+    try {
+      const question = req.body;
```
Comment (cristim67) [inline server/app/question/question.router.ts:44]: check is null

File: server/app/question/question.router.ts (hunk 9 above)
Comment (cristim67) [inline server/app/question/question.router.ts:49]: console.log the error.

File: server/app/question/question.router.ts [hunk 11]
```diff
@@ -0,0 +1,85 @@
...
+    const id: number = parseInt(req.params.id, 10);
+    try {
+      const question = req.body;
+      const updatedQuestion = await QuestionService.updateQuestion(
+        question,
+        id
+      );
+      return res.status(200).json(updatedQuestion);
+    } catch (error: any) {
+      return res.status(500).json({ errors: error.message });
```
Comment (cristim67) [inline server/app/question/question.router.ts:71]: console.log the error message.

File: server/app/question/question.router.ts [hunk 12]
```diff
@@ -0,0 +1,85 @@
...
+  "/:id",
+  body("question").isString(),
+  async (req: Request, res: Response) => {
+    const errors = validationResult(request);
+    if (!errors.isEmpty()) {
+      return response.status(400).json({ errors: errors.array });
+    }
+    const id: number = parseInt(req.params.id, 10);
+    try {
+      const question = req.body;
```
Comment (cristim67) [inline server/app/question/question.router.ts:64]: check is null

File: server/app/question/question.router.ts [hunk 13]
```diff
@@ -0,0 +1,85 @@
...
+);
+
+// DELETE by id
+questionRouter.delete("/:id", async (req: Request, res: Response) => {
+  const id: number = parseInt(req.params.id, 10);
+  try {
+    await QuestionService.deleteQuestion(id);
+    return res.status(200).json({ message: "Question has been deleted" });
+  } catch (error: any) {
+    return res.status(500).json({ errors: error.message });
```
Comment (cristim67) [inline server/app/question/question.router.ts:83]: console.log the error message.

File: server/app/question/question.router.ts [hunk 14]
```diff
@@ -0,0 +1,85 @@
...
+    }
+  }
+);
+
+// DELETE by id
+questionRouter.delete("/:id", async (req: Request, res: Response) => {
+  const id: number = parseInt(req.params.id, 10);
+  try {
+    await QuestionService.deleteQuestion(id);
+    return res.status(200).json({ message: "Question has been deleted" });
```
Comment (cristim67) [inline server/app/question/question.router.ts:81]: check is deleted.

File: server/app/question/question.service.ts [hunk 15]
```diff
@@ -0,0 +1,102 @@
+import { Prisma, PrismaClient } from "@prisma/client";
+
+const db = new PrismaClient();
```
Comment (cristim67) [inline server/app/question/question.service.ts:3]: move this in `models` folder

File: server/app/question/question.service.ts [hunk 16]
```diff
@@ -0,0 +1,102 @@
...
+type Question = {
+  id: number;
+  question: string;
+  answers: Prisma.JsonValue;
+  createdAt: Date;
+  updatedAt: Date;
+};
+
+// List all questions
+export const listQuestions = async (): Promise<Question[]> => {
```
Comment (cristim67) [inline server/app/question/question.service.ts:14]: i think for all this you need a try catch block

File: server/prisma/schema.prisma [hunk 17]
```diff
@@ -0,0 +1,29 @@
+// This is your Prisma schema file,
+// learn more about it in the docs: https://pris.ly/d/prisma-schema
+
+// Looking for ways to speed up your queries, or scale easily with your serverless or edge functions?
+// Try Prisma Accelerate: https://pris.ly/cli/accelerate-init
+
+generator client {
+  provider = "prisma-client-js"
+}
```
Comment (cristim67) [inline server/prisma/schema.prisma:9]: here you need to add `binaryTargets` param

Comment (MDGcode) [reply server/prisma/schema.prisma:9]: Done

File: server/app/question/question.service.ts (hunk 16 above)
Comment (MDGcode) [reply server/app/question/question.service.ts:14]: Done

File: server/controllers/question/questionController.ts [hunk 18]
```diff
@@ -0,0 +1,125 @@
+import express from "express";
+import type { Request, Response } from "express";
+import { body, validationResult } from "express-validator";
+import { Prisma } from "@prisma/client";
+import * as QuestionService from "../../services/question/question.service";
```
Comment (cristim67) [inline server/controllers/question/questionController.ts:5]: here you can use import like as alias with `@` but it`s not mandatory.

you can check more at this post https://stackoverflow.com/questions/70705095/how-to-import-module-using-symbol-nodeexpresstypescript

File: server/prisma/schema.prisma [hunk 19]
```diff
@@ -0,0 +1,30 @@
+// This is your Prisma schema file,
+// learn more about it in the docs: https://pris.ly/d/prisma-schema
+
+// Looking for ways to speed up your queries, or scale easily with your serverless or edge functions?
+// Try Prisma Accelerate: https://pris.ly/cli/accelerate-init
+
+generator client {
+  provider      = "prisma-client-js"
+  binaryTargets = ["native", "linux-musl", "darwin", "windows"]
```
Comment (cristim67) [inline server/prisma/schema.prisma:9]: here you need to add only the target that match the deploy server, none of this values match the deploy server.

here you need to add "linux-arm64-openssl-3.0.x" or "linux-arm64-openssl-1.0.x"

File: server/utils/db.server.ts [hunk 20]
```diff
@@ -0,0 +1,15 @@
...
+  var __db: PrismaClient | undefined;
+}
+
+if (!global.__db) {
+  global.__db = new PrismaClient();
+}
+
+db = global.__db;
+
+export { db };
```
Comment (cristim67) [inline server/utils/db.server.ts:15]: that`s not utils

**Review by cristim67 — APPROVED** [review]
looks good, you can merge after

**Pull request merged**